lfsapi/auth.go

```diff
@@ -24,6 +24,19 @@ var (
 // authentication from netrc or git's credential helpers if necessary,
 // supporting basic and ntlm authentication.
 func (c *Client) DoWithAuth(remote string, req *http.Request) (*http.Response, error) {
+	req.Header = c.extraHeadersFor(req)
```

Comment:
The Client type is supposed to have `Do()` and `DoWithAuth()` as the two main ways to make a request. However, `DoWithAuth()` eventually calls `Do()` in `doWithCreds()`:

https://github.com/git-lfs/git-lfs/blob/5427600232727855328b7c40021b5bb20a5424b4/lfsapi/auth.go#L58-L63

One easy fix is to pull out a `do()` function from `Do()`:

https://github.com/git-lfs/git-lfs/blob/5427600232727855328b7c40021b5bb20a5424b4/lfsapi/client.go#L86-L94

Then, make sure any `lfsapi.Client` funcs that call `Do()`, call this new `do()` instead. Does that make sense?

There's likely a better extraction hiding in here too. I don't like having `Do()` and `DoWithAuth()`. But this would require a lot more thought and work, where a quick fix might be better for the extra headers fix.

Reply:
Great call! I think we might even not have to deal with this based on your comment in https://github.com/git-lfs/git-lfs/pull/2733#discussion_r151534714, since the point of this `if` was to avoid calling `getCreds()`, since my original assumption was that `getCreds()` _always_ contacted the credential helper.

Since it double checks for authorization, I think the fix here is only ensuring that we call `c.extraHeadersFor`.

Reply:
> since my original assumption was that getCreds() always contacted the credential helper.

It used to, but I fixed it: https://github.com/git-lfs/git-lfs/pull/2695

> Since it double checks for authorization, I think the fix here is only ensuring that we call c.extraHeadersFor.

Yeah, I suppose `extraHeadersFor()` could be called in `DoWithAuth()`, and again when `Do()` is called by `doWithCreds()`. It is rebuilding the `http.Header`, but should _retain_ the existing header without causing problems. 

Though I think if `DoWithAuth()` somehow changes one of the headers that is one of the git-configured extra headers, and is not `Authorization`, then the changes would be clobbered when `extraHeadersFor()` is called again inside `Do()`.

Reply:
@technoweenie -- ah, I think I understand now, and thanks for (double) explaining this to me! I agree that some larger changes may be useful to think about here, but for now I opted for `do()`, which handles redirects and errors, but does _not_ add extra headers.

After changing the behavior of `lfsapi` to no longer itself call `Do()`, I believe that this is ready to go.